### PR TITLE
Complete rewrite

### PR DESCRIPTION
**Pull request by dissolve**
Not sure who would be most involved, but I have basically rewritten the entire code base from scratch.
Almost everything is supported, the only tests failing are ones for the include-pattern, one that nokogiri seems to change the html when modifying hrefs (because of the relative url), and the rest I believe the test suite has incorrect results.

This is a breaking change to how people would be using it.  Happy to discuss any aspect of it here or on IRC.

### CODE REVIEW

**Comment by veganstraightedge** [comment]
I'm digging into this now. I'll be back soon with real thoughts.

**Comment by jeena** [comment]
So you mean this breaks the API? In which way?

**Comment by dissolve** [comment]
Going to make some clearer listing of what has changed.  But the functions all, first, last, format, and any plurals would not work.  This branch wouldn't be merged until after a new release that would add deprecation warnings for all of them.

In addition the classes are all simply a ParserResult class, so if anyone checks for class name, there are possible issues

File: lib/microformats2/format_parser.rb [hunk 1]
```diff
@@ -1,60 +1,346 @@
...
+
+            if element.name == 'img' and not element.attribute('alt').nil?
+              @properties['name'] = [element.attribute('alt').value.strip]
+            elsif element.name == 'area' and not element.attribute('alt').nil?
+              @properties['name'] = [element.attribute('alt').value.strip]
+            elsif element.name == 'abbr' and not element.attribute('title').nil?
+              @properties['name'] = [element.attribute('title').value.strip]
+
+            else
+                child_nodes = element.children.select{|n| not n.is_a?(Nokogiri::XML::Text)}
```
Comment (dissolve) [inline lib/microformats2/format_parser.rb]: Yes I had noticed this while writing it but was waiting until done to go back and fix it all.  I have vim always set to 4 spaces, but will definitely fix this

File: lib/microformats2.rb [hunk 2]
```diff
@@ -1,30 +1,20 @@
...
-require "microformats2/implied_property/name"
-require "microformats2/implied_property/photo"
-require "microformats2/implied_property/url"
+require "microformats2/time_property_parser"
 
 module Microformats2
   class << self
-    def parse(html)
-      Parser.new.parse(html)
+    def parse(html, base=nil)
```
Comment (dissolve) [inline lib/microformats2.rb]: Agreed, will make this change

File: lib/microformats2/absolute_uri.rb [hunk 3]
```diff
@@ -5,10 +5,14 @@ class AbsoluteUri
     def initialize(base, relative)
       @base = base
       @relative = relative
+      @base = @base.strip unless @base.nil?
```
Comment (dissolve) [inline lib/microformats2/absolute_uri.rb]: Huh, learn something new every day.

**Comment by dissolve** [comment]
thinking its best to move this to a new branch on the main project (after i clean it up a bit)

**Comment by dissolve** [comment]
applied the changes you suggested @veganstraightedge
Closing the PR and created a new 3.0.0 branch to continue work there before any sort of final PR is made against master

**Pull request closed without merging**